### ISSUE TITLE
Standing Queries are not considered when calculating the order of indexes when combined with normal clauses

### ISSUE THREAD

**Issue by plenderyou** [npgall/cqengine#232]
When calculating the order of Result Sets to process for a logical query, standing indexes are not included.

I put this code into the CollectionQueryEngine

`   <A> ResultSet<O> getResultSetWithLowestRetrievalCost(SimpleQuery<O, A> query, QueryOptions queryOptions) {
        // First check if a UniqueIndex is available, as this will have the lowest cost...
        Index<O> uniqueIndex = uniqueIndexes.get(query.getAttribute());
        if (uniqueIndex!= null && uniqueIndex.supportsQuery(query, queryOptions)){
            return uniqueIndex.retrieve(query, queryOptions);
        }

        ResultSet<O> lowestCostResultSet = null;
        int lowestRetrievalCost = 0;
        final Index<O> oIndex = standingQueryIndexes.get(query);
        if( oIndex != null ) {
            if( oIndex.supportsQuery(query, queryOptions) ) {
                ResultSet<O> thisIndexResultSet = oIndex.retrieve(query, queryOptions);
                lowestCostResultSet = thisIndexResultSet;
                lowestRetrievalCost = thisIndexResultSet.getRetrievalCost();
            }
        }

        // Examine other (non-unique) indexes...
        Iterable<Index<O>> indexesOnAttribute = getIndexesOnAttribute(query.getAttribute());

        // Choose the index with the lowest retrieval cost for this query...
        for (Index<O> index : indexesOnAttribute) {
            if (index.supportsQuery(query, queryOptions)) {
                ResultSet<O> thisIndexResultSet = index.retrieve(query, queryOptions);
                int thisIndexRetrievalCost = thisIndexResultSet.getRetrievalCost();
                if (lowestCostResultSet == null || thisIndexRetrievalCost < lowestRetrievalCost) {
                    lowestCostResultSet = thisIndexResultSet;
                    lowestRetrievalCost = thisIndexRetrievalCost;
                }
            }
        }

        if (lowestCostResultSet == null) {
            // This should never happen (would indicate a bug);
            // the fallback index should have been selected in worst case...
            throw new IllegalStateException("Failed to locate an index supporting query: " + query);
        }
        return new CostCachingResultSet<O>(lowestCostResultSet);
    }`

And they were but I'm not sure of the implications

**Comment by npgall** [comment]
Are you facing a particular problem with this? I don't currently see a problem in the code, but please let me know if I'm missing something.

A standing query index can be added on a logical query or on a simple query.

The [retrieveRecursive()](https://github.com/npgall/cqengine/blob/3.2.0/code/src/main/java/com/googlecode/cqengine/engine/CollectionQueryEngine.java#L891) method evaluates all queries - both logical and simple queries. As far as I can see, the first thing it does is try to use any standing query indexes if those are available to answer the query. 

It does subsequently call [getResultSetWithLowestRetrievalCost()](https://github.com/npgall/cqengine/blob/3.2.0/code/src/main/java/com/googlecode/cqengine/engine/CollectionQueryEngine.java#L307), but that's only for simple queries and (AFAICS) only called if standing query indexes are not available.

**Comment by plenderyou** [comment]
The problem is that when a logical query is evaluated, it calculates the retrieval cost for each index and then uses that order to calculate the order of evaluation, defaulting to FallbackIndex (for standing queries).

I've attached a reproducer and the 'HacK' I put into the CollectionQueryEngine code.



[standingQueryStuff.zip](https://github.com/npgall/cqengine/files/3196838/standingQueryStuff.zip)

**Comment by plenderyou** [comment]
Did you have a look at the reproducer I attached

**Comment by npgall** [comment]
Hi @plenderyou ,

Thanks for supplying the reproducer! I was able to reproduce it using that.

It seems that standing query indexes indeed were not being considered for nested simple queries, they were only considered for nested logical queries.

I've just pushed a fix which resolves this, and also some additional tests for the different scenarios.

Thanks for your help to find this!

Best regards,
Niall
